### PR TITLE
Improve the handling of AMQP Client Options customizers

### PR DESCRIPTION
**Pull request by cescoffier**
See https://github.com/smallrye/smallrye-reactive-messaging/issues/1873

Not totally happy, but at least better and fix the behavior reported in #1873.

### CODE REVIEW

File: smallrye-reactive-messaging-amqp/src/main/java/io/smallrye/reactive/messaging/amqp/AmqpClientHelper.java [hunk 1]
```diff
@@ -51,10 +51,16 @@ static AmqpClient createClientFromClientOptionsBean(Vertx vertx, Instance<AmqpCl
...
-        mergeTo(clientOptions, completeOptions);
-        return AmqpClient.create(vertx, completeOptions);
+        AmqpClientOptions channelOptions = getOptionsFromChannel(config);
+        AmqpClientOptions customizerOptions = options.get();
+        merge(channelOptions, customizerOptions, config);
+        return AmqpClient.create(vertx, customizerOptions);
+    }
+
+    static boolean isSetInChannelConfiguration(String attribute, AmqpConnectorCommonConfiguration configuration) {
+        System.out.println(
```
Comment (ozangunalp) [inline smallrye-reactive-messaging-amqp/src/main/java/io/smallrye/reactive/messaging/amqp/AmqpClientHelper.java:61]: Remove or replace this with a log

File: smallrye-reactive-messaging-amqp/src/main/java/io/smallrye/reactive/messaging/amqp/AmqpClientHelper.java [hunk 2]
```diff
@@ -63,44 +69,67 @@ static AmqpClient createClientFromClientOptionsBean(Vertx vertx, Instance<AmqpCl
...
         }
-        if (useSsl) {
+
+        // custom.getPort() will return a value (might be the default port, or not, or the default port as set by the user)
+        // the channel port may be set (using the alias or connector config)
+        // we can compare, but we cannot decide which one is going to be used.
+        // so, we use the one from the customizer except is explicitly set on the channel.
+        // The same apply for ssl, reconnect-attempts, reconnect-interval and connect-timeout.
+
+        if (isSetInChannelConfiguration("ssl", config)) {
```
Comment (ozangunalp) [inline smallrye-reactive-messaging-amqp/src/main/java/io/smallrye/reactive/messaging/amqp/AmqpClientHelper.java:117]: Isn't this property `use-ssl` ?

File: smallrye-reactive-messaging-amqp/src/main/java/io/smallrye/reactive/messaging/amqp/AmqpClientHelper.java [hunk 3]
```diff
@@ -63,44 +69,67 @@ static AmqpClient createClientFromClientOptionsBean(Vertx vertx, Instance<AmqpCl
...
-        int port = channel.getPort();
-        boolean useSsl = channel.isSsl();
-        int reconnectAttempts = channel.getReconnectAttempts();
-        long reconnectInterval = channel.getReconnectInterval();
-        int connectTimeout = channel.getConnectTimeout();
-
-        if (username != null) {
+    static void merge(AmqpClientOptions channel, AmqpClientOptions custom, AmqpConnectorCommonConfiguration config) {
+        String usernameFromChannelConfig = channel.getUsername();
+        String passwordFromChannelConfig = channel.getPassword();
```
Comment (ozangunalp) [inline smallrye-reactive-messaging-amqp/src/main/java/io/smallrye/reactive/messaging/amqp/AmqpClientHelper.java:76]: These are not used anymore ?

File: smallrye-reactive-messaging-amqp/src/main/java/io/smallrye/reactive/messaging/amqp/AmqpClientHelper.java [hunk 4]
```diff
@@ -63,44 +69,67 @@ static AmqpClient createClientFromClientOptionsBean(Vertx vertx, Instance<AmqpCl
...
-        String password = channel.getPassword();
-        String host = channel.getHost();
-        int port = channel.getPort();
-        boolean useSsl = channel.isSsl();
-        int reconnectAttempts = channel.getReconnectAttempts();
-        long reconnectInterval = channel.getReconnectInterval();
-        int connectTimeout = channel.getConnectTimeout();
-
-        if (username != null) {
+    static void merge(AmqpClientOptions channel, AmqpClientOptions custom, AmqpConnectorCommonConfiguration config) {
```
Comment (ozangunalp) [inline smallrye-reactive-messaging-amqp/src/main/java/io/smallrye/reactive/messaging/amqp/AmqpClientHelper.java:74]: I don't completely follow what is going on in this method. Maybe if we move `getOptionsFromChannel` inside the `merge` method and return the merged options, it'll be easier to follow.

**Review by ozangunalp — CHANGES_REQUESTED** [review]

File: smallrye-reactive-messaging-amqp/src/main/java/io/smallrye/reactive/messaging/amqp/AmqpClientHelper.java (hunk 3 above)
Comment (cescoffier) [reply smallrye-reactive-messaging-amqp/src/main/java/io/smallrye/reactive/messaging/amqp/AmqpClientHelper.java:76]: removed

File: smallrye-reactive-messaging-amqp/src/main/java/io/smallrye/reactive/messaging/amqp/AmqpClientHelper.java (hunk 4 above)
Comment (cescoffier) [reply smallrye-reactive-messaging-amqp/src/main/java/io/smallrye/reactive/messaging/amqp/AmqpClientHelper.java:74]: done.

**Review by ozangunalp — APPROVED** [review]

**Pull request merged**